[#42] Allow users to select a sheet

kndungu opened the pull request:
**What's New**

Upon clicking on `Add File` the user will now be greeted by two options:

1. Selection (the default) -> Binds the currently selected range.
2. Sheet -> Binds the worksheet's first 150 columns (A to ET) and 10000 rows (can be changed, the worksheet has over 10 million rows).

**Update to Logic**

The create binding function now accepts as its argument an object containing:

1. The file name entered by the user, this also serves as the named item's name.
2. The sheet where the binding is being made.
3. The range, will be the user's selection if `Selection` was selected and `A1:ET10000` if `Sheet` was selected.

Create binding first creates a [nameditem](https://dev.office.com/reference/add-ins/excel/nameditemcollection) using the passed in arguments. A binding is then created from the nameditem using the [addFromNamedItemAsync](https://dev.office.com/reference/add-ins/shared/bindings.addfromnameditemasync) function

alevicki commented on src/components/AddDataModal.js:
```diff
@@ -100,7 +108,9 @@ class AddDataModal extends Component {
     } else {
-      this.submitBinding();
+      const sheet = `Sheet${this.getSheetNumber(this.props.range)}`;
+      const range = this.props.range ? this.props.range.substring(this.props.range.indexOf("!") + 1) : ''; 
```
Use single quotes for `indexOf('!')`

alevicki commented on src/components/AddDataModal.js:
```diff
@@ -129,6 +139,62 @@ class AddDataModal extends Component {
+      let i;
+      let j;
+
+
```
Remove extra newline

alevicki commented on src/components/AddDataModal.js:
```diff
@@ -129,6 +139,62 @@ class AddDataModal extends Component {
     this.props.close();
   }
 
+  rangetoRowsCols = (range = "") => {
```
Instead of doing this, you should be able to return the range object instead of the address from the `getCurrentlySelectedRange` That way you can use the `rowCount` and `columnCount` to generate this string.

alevicki commented on src/components/AddDataModal.js:
```diff
@@ -149,25 +215,40 @@ class AddDataModal extends Component {
-                Select the area to bind in the worksheet, it will be reflected here.
+                {
+                  this.state.selectSheet ?
+                  "Select the sheet you'd like to add, it will be refelected here." :
```
Single quotes

alevicki commented on src/components/AddDataModal.js:
```diff
@@ -149,25 +215,40 @@ class AddDataModal extends Component {
+              </div>
+              <div className='selection'>
+                <label className='radio-button'>
+                  <input type="radio" name="selection" value='sheet' />
```
Single quotes

alevicki commented on src/components/AddDataModal.js:
```diff
@@ -149,25 +215,40 @@ class AddDataModal extends Component {
+            <form className="selection-form" onClick={this.changeSelection}>
+              <div className='selection'>
+                <label className='radio-button'>
+                  <input type="radio" name="selection" value="selection" defaultChecked />
```
Single quotes

alevicki commented on src/components/AddDataModal.js:
```diff
@@ -149,25 +215,40 @@ class AddDataModal extends Component {
-                  type='text' />
-              </InputGroup>
+            {excelApiSupported &&
+            <form className="selection-form" onClick={this.changeSelection}>
```
Single quotes

alevicki commented on src/components/AddDataModal.js:
```diff
@@ -149,25 +215,40 @@ class AddDataModal extends Component {
               <HelpBlock>
-                Select the area to bind in the worksheet, it will be reflected here.
+                {
+                  this.state.selectSheet ?
```
Pull `selectSheet` out in the `state` destructuring at the top of the `render` function

alevicki commented on src/components/AddDataModal.js:
```diff
@@ -129,6 +139,62 @@ class AddDataModal extends Component {
+    return `${numberOfRows} Rows x ${numberOfColumns} Columns`;
+  }
+
+  getSheetNumber = (range = "") => {
```
You can also pull the worksheet out with the range if you want to.  This one is pretty simple, so it is fine to stay as is.

alevicki commented on src/App.js:
```diff
@@ -166,9 +166,13 @@ class App extends Component {
+    await this.office.createSelectionRange(selection.name, selection.sheet, selection.range)
+
+    // Temporary workaround: The last namedItem added is not visible so we have to add twice 
+    await this.office.createSelectionRange(selection.name + "Temp", selection.sheet, selection.range)
```
Let's investigate this a bit further to see if we can't figure out what is going on.  It doesn't seem to make sense that you would need to call this twice.  There is probably a refresh names somewhere.

alevicki commented on src/OfficeConnector.js:
```diff
@@ -95,17 +95,38 @@ export default class OfficeConnector {
     });
   }
 
+  createBinding(name, options) {
```
Remove `options` since it appears to be unused.

alevicki commented:
In trying to bind to a selection, I get a "The named item does not exist." error in the UI.

Also, did you discuss the decision to not actually bind to the entire worksheet with @rflprr?

kndungu commented:
That's the error the double creation of the nameditem is supposed to address. Investigating.

Of the thousands of columns in the sheet we can only bind to 150 so that only leaves the number of rows we should bind. 10,000 is a stand in number before @rflprr gives me a figure.

alevicki commented:
@kndungu There seems to be an issue where when you bind to an area then refresh the page, it is not aware of when you make changes to selected areas.  In addition, clicking on Save Files does not sync up the changes to the server.

kndungu commented:
@alevicki 

**The named item does not exist error**
Fixed in commits b42a138 and c63d4ce

**Not binding entire worksheet**
Fixed in 6f42318. All rows are now bound when a user selects the sheet. The binding is still restricted to the first 150 columns as per the current limit (https://github.com/datadotworld/excel-add-in/blob/5bbe3d1e0395e344fdfab1ec7dd49dcb920def11/src/App.js#L172).

**Losing binding upon page refresh**
Confirmed; able to reproduce. This is however the current behaviour and is not due to the new changes. I'll create a separate issue to address it.

**Clicking on Save Files does not sync up the changes to the server.**
Unable to reproduce, you may have stumbled on an edge case. Please share steps for reproduction.

The PR only updates how bindings are created and not how they are handled after creation, the bug may be in the current code in which case we'd need to create a separate issue.

kndungu commented:
@rflprr 

> The idea for this feature was to create a dynamic binding, but the way you implemented it, it is still static.

Excel's API does not allow us to bind to entire sheets so we have to artificially create this functionality using ranges.

When a user binds to a sheet they are basically saying that they want to be able to enter data anywhere within the sheet and have Excel detect the change. Accomplishing this with ranges entails binding to the range `A1:ET1048576`

When the user syncs the sheet to data.world, empty columns and rows are ignored. So if the user enters data into 5 rows and 3 columns this 5 x 3 grid is what will appear on data.world even though to Excel the binding is a 150 x 1048576 grid.

If 2 rows and 4 columns  are later added Excel will detect the change because the new rows and columns are within the 150 x 1048576 bound grid. Upon syncing, a 7 x 7 grid will appear on data.world.

This is what makes it dynamic even though the actual binding is static

kndungu commented:
> when I bind to a sheet, and later edit the binding, the add-in shows that the file is no longer bound to a sheet, but instead to a selection (range).

This is because the underlying implementation of sheet binding still relies on ranges.

This is a UI change which we had overlooked, will change the text to be `SheetX` when the binding is to the sheet instead of the range.

kndungu commented:
> when I look at the CSV that we upload to data.world there are lots of empty cells that are uploaded unnecessarily.

The add-in uploads the grid that contains all cells with values and ignores everything outside this grid

On data.world empty rows are ignored but empty columns are not.

So 

<img width="743" alt="screen shot 2017-12-08 at 06 18 37" src="https://user-images.githubusercontent.com/17295379/33752070-ab2e9140-dbdf-11e7-8b50-c3aa6aea33dd.png">

Is uploaded as

<img width="815" alt="screen shot 2017-12-08 at 06 19 19" src="https://user-images.githubusercontent.com/17295379/33752088-c767e3b6-dbdf-11e7-8bae-0d5a5dd9b339.png">

alevicki commented on src/App.js:
```diff
@@ -371,7 +421,17 @@ class App extends Component {
 
     if (binding) {
-      this.office.select(binding.rangeAddress);
+      const maxRows = 1048576;
```
You use these numbers in several places, they should be moved to a constants file.

alevicki commented on src/components/BindingListItem.js:
```diff
@@ -62,6 +62,24 @@ class BindingsListItem extends Component {
+      const maxColumns = 150;
+
+      // If it's a sheet binding just show the sheet number
+      if (binding.rowCount === maxRows && binding.columnCount === maxColumns) {
```
This test could also be moved to a centralized function somewhere.

alevicki commented:
@kndungu My requested changes aren't blocking, they can be updated after merge.
